tests/shippers/test_logstash.py

```diff
@@ -0,0 +1,193 @@
+            expected_event["_id"] = "_id"
+            assert _payload == [expected_event, expected_event]
+
+            return 200, {}, "okay"
+
+        def event_id_generator(event: dict[str, Any]) -> str:
+            return "_id"
+
+        url = "http://logstash_url"
+        responses.add_callback(responses.PUT, url, callback=request_callback)
+        logstash_shipper = LogstashShipper(logstash_url=url, max_batch_size=2)
+        logstash_shipper.set_event_id_generator(event_id_generator)
+        logstash_shipper.send(_dummy_event)
+        logstash_shipper.send(_dummy_event)
+
+    @responses.activate
+    def test_send_failures(self) -> None:
+        url = "http://logstash_url"
+        with self.subTest("Does not exceed max_retries"):
+            responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            responses.put(url=url, status=200)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+        with self.subTest("Exceeds max retries, replay handler set"):
+            for i in range(_MAX_RETRIES):
+                responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            replay_handler = MagicMock(side_effect=_dummy_replay_handler)
+            logstash_shipper.set_replay_handler(replay_handler)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+            replay_handler.assert_called_once_with("logstash", {}, _dummy_event)
+        with self.subTest("Exceeds max retries, replay handler not set"):
+            for i in range(_MAX_RETRIES):
+                responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            replay_handler = MagicMock(side_effect=_dummy_replay_handler)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+            replay_handler.assert_not_called()
+        with self.subTest("Authentication error, request is not retried"):
+            responses.put(url=url, status=401)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            replay_handler = MagicMock(side_effect=_dummy_replay_handler)
+            logstash_shipper.set_replay_handler(replay_handler)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+            replay_handler.assert_called_once_with("logstash", {}, _dummy_event)
+
+    def test_send_https_ssl_fingerprint(self) -> None:
+        certpath = os.path.join(os.path.dirname(__file__), "ssl", "localhost.crt")
+        keypath = os.path.join(os.path.dirname(__file__), "ssl", "localhost.pkcs8.key")
+        server_address = ("localhost", 8080)
+        httpd = http.server.HTTPServer(server_address, http.server.SimpleHTTPRequestHandler)
```

Comment:
I'd rather use the real logstash from docker, in an integration test, also we should test where the case providing a non matching ssl fingerprint fill the replay queue,

Reply:
Yep, I agree. These were added when we did not have integration tests but they do not make a lot of sense now.